File: Client/Captions/CaptionsLib/source/BaseCaptionReader.vb
'Filename: BaseCaptionReader.vb
'Version: 20121016

Imports ClipFlair.CaptionsLib.models

Imports System.IO
Imports System.Text
Imports Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions

Namespace ClipFlair.CaptionsLib

  Public MustInherit Class BaseCaptionReader
    Implements ICaptionsReader

#Region "Methods"

    Public Overloads Sub ReadCaptions(ByVal captions As CaptionRegion, ByVal path As String, ByVal theEncoding As Encoding) Implements ICaptionsReader.ReadCaptions
      'not clearing any existing captions, just appending to the end (the CaptionRegion object can choose whether it will sort the Captions by start time or not after the appending)
      Using reader As New StreamReader(path, theEncoding, True) '"Using" makes sure the resource is immediately deallocated at "End Using"
        ReadCaptions(captions, reader)
      End Using
    End Sub

    Public Overloads Sub ReadCaptions(ByVal captions As CaptionRegion, ByVal stream As Stream, ByVal theEncoding As Encoding) Implements ICaptionsReader.ReadCaptions
      'not clearing any existing captions, just appending to the end (the CaptionRegion object can choose whether it will sort the Captions by start time or not after the appending)
      Using reader As New StreamReader(stream, theEncoding, True) '"Using" makes sure the resource is immediately deallocated at "End Using"
        ReadCaptions(captions, reader)
      End Using
    End Sub

    Public Overloads Sub ReadCaptions(ByVal captions As CaptionRegion, ByVal reader As TextReader) Implements ICaptionsReader.ReadCaptions
      Try
        ReadHeader(reader)
        While reader.Peek <> -1
          Dim caption As New CaptionElement()
          ReadCaption(caption, reader)
          captions.Children.Add(caption)
        End While
        ReadFooter(reader)
      Finally
        reader.Close()
      End Try
    End Sub

    Public Overridable Sub ReadHeader(ByVal reader As TextReader) Implements ICaptionsReader.ReadHeader
      'can override at descendents
    End Sub

    Public MustOverride Sub ReadCaption(ByVal caption As CaptionElement, ByVal reader As TextReader) Implements ICaptionsReader.ReadCaption

    Public Overridable Sub ReadFooter(ByVal reader As TextReader) Implements ICaptionsReader.ReadFooter
      'can override at descendents
    End Sub

#End Region

  End Class

End Namespace

File: Client/Captions/CaptionsLib/source/BaseCaptionReader.cs
//Project: ClipFlair (http://ClipFlair.codeplex.com)
//Filename: BaseCaptionReader.cs
//Version: 20131105

using ClipFlair.CaptionsLib.Models;
using Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions;
using System.IO;
using System.Text;

namespace ClipFlair.CaptionsLib
{

	public abstract class BaseCaptionReader : ICaptionsReader
	{

		#region --- Methods ---

		public void ReadCaptions<T>(CaptionRegion captions, string path, Encoding theEncoding) where T : CaptionElement, new()
		{
			//not clearing any existing captions, just appending to the end (the CaptionRegion object can choose whether it will sort the Captions by start time or not after the appending)
			using (StreamReader reader = new StreamReader(path, theEncoding, true)) {
				//the Using statement will close the file created when finished
				ReadCaptions<T>(captions, reader);
			}
		}

		public void ReadCaptions<T>(CaptionRegion captions, Stream stream, Encoding theEncoding) where T : CaptionElement, new()
		{
			//not clearing any existing captions, just appending to the end (the CaptionRegion object can choose whether it will sort the Captions by start time or not after the appending)
			StreamReader reader = new StreamReader(stream, theEncoding, true);
			//no Using statement, do not close the stream when finished
			ReadCaptions<T>(captions, reader);
		}

		public void ReadCaptions<T>(CaptionRegion captions, TextReader reader) where T : CaptionElement, new()
		{
			ReadHeader(reader);
			while (reader.Peek() != -1) {
				T caption = new T();
				ReadCaption(caption, reader);
				captions.Children.Add(caption);
			}
			ReadFooter(reader);
		}

		public virtual void ReadHeader(TextReader reader)
		{
			//can override at descendents
		}

		public abstract void ReadCaption(CaptionElement caption, TextReader reader);

		public virtual void ReadFooter(TextReader reader)
		{
			//can override at descendents
		}

		#endregion

	}

}